mobile/lib/modules/home/ui/asset_grid/asset_grid_data_structure.dart

```diff
@@ -33,85 +35,104 @@ class RenderAssetGridElement {
   });
 }
 
-List<RenderAssetGridElement> assetsToRenderList(
-  List<Asset> assets,
-  int assetsPerRow,
-) {
-  List<RenderAssetGridElement> elements = [];
-
-  int cursor = 0;
-  while (cursor < assets.length) {
-    int rowElements = min(assets.length - cursor, assetsPerRow);
-    final date = assets[cursor].createdAt;
-
-    final rowElement = RenderAssetGridElement(
-      RenderAssetGridElementType.assetRow,
-      date: date,
-      assetRow: RenderAssetGridRow(
-        assets.sublist(cursor, cursor + rowElements),
-      ),
-    );
-
-    elements.add(rowElement);
-    cursor += rowElements;
-  }
-
-  return elements;
-}
-
-List<RenderAssetGridElement> assetGroupsToRenderList(
+/// Converts assets grouped by date to the data model required by Immich's custom asset grid implementation.
+Future<List<RenderAssetGridElement>> assetGroupsToRenderList(
   Map<String, List<Asset>> assetGroups,
   int assetsPerRow,
-) {
-  List<RenderAssetGridElement> elements = [];
-  DateTime? lastDate;
-
-  assetGroups.forEach((groupName, assets) {
-    try {
-      final date = DateTime.parse(groupName);
+) async {
+
+  processData(Map<String, Object> data) {
+    final monthFormat = DateFormat(data["month_format"]! as String);
+    final dayFormatSameYear = DateFormat(data["day_format"]! as String);
+    final dayFormatOtherYear = DateFormat(data["day_format_year"]! as String);
+    final groups = data["groups"]! as Map<String, List<Asset>>;
+    final perRow = data["per_row"]! as int;
+
+    List<RenderAssetGridElement> elements = [];
+    DateTime? lastDate;
+
+    groups.forEach((groupName, assets) {
+      try {
+        final date = DateTime.parse(groupName);
+
+        if (lastDate == null || lastDate!.month != date.month) {
+          // Month title
+
+          var monthTitleText = groupName;
+
+          try {
+            monthTitleText = monthFormat.format(DateTime.parse(groupName));
+          } catch (e) {
+            log.severe("Failed to format date for month title: $groupName");
+          }
```

Comment:
I believe there is a non throwing version of this method that would prevent using a try/catch. Try/catch is usually slower.